Auth cleanup

Vielheim opened the pull request:
Changed Authentication to Pub-Sub architecture
- Create AuthenticationVM
- Remove reference to FirebaseAuth.auth() in service and repo layer (Reference through AuthenticationManager)

Commit:
Convert Authentication to Pub-Sub design

Added AuthenticationViewModel (Singleton)
Added AuthenticationService to AppConfig (for Dependency Injection)
Shifted listener logic to AuthenticationManager

Commit:
Remove reference to FirebaseAuth in Service Layer

Commit:
Remove reference to FirebaseAuth in repository

Commit:
Add protection to AuthVM

keanecjy commented on TourMate/TourMate/Frontend/User/Services/FirebaseUserService.swift:
```diff
@@ -6,15 +6,15 @@
 
     func addUser(_ user: User) async -> (Bool, String) {
-        guard let currentUser = Auth.auth().currentUser,
+        guard let currentUser = authenticationManager.getCurrentFirebaseUser(),
```
I think its better if we access the firebaseuserservice instead to check whether the user is logged in, and let the auth service be the only one accessing the auth manager.

keanecjy commented on TourMate/TourMate/Frontend/Authentication/Services/FirebaseAuthenticationService.swift:
```diff
@@ -9,44 +9,40 @@ import Foundation
+final class FirebaseAuthenticationService: AuthenticationService {
 
-    private(set) static var shared = FirebaseAuthenticationService()
+    @Published private(set) var userIsLoggedIn = false
```
This does not seem used

keanecjy commented on TourMate/TourMate/Backend/Persistence/Repository/FirebaseRepository.swift:
```diff
@@ -22,7 +23,7 @@ class FirebaseRepository: Repository {
     @MainActor
     func addItem<T: FirebaseAdaptedData>(id: String, item: T) async -> (hasAddedItem: Bool, errorMessage: String) {
-        guard Auth.auth().currentUser != nil else {
+        guard authenticationManager.getCurrentFirebaseUser() != nil else {
```
Perhaps we can use a function like userIsLoggedIn? I think the logic whether the user is logged in should be contained within the auth manager

keanecjy commented on TourMate/TourMate/Backend/Authentication/Managers/FirebaseAuthenticationManager.swift:
```diff
@@ -10,25 +10,45 @@ import Firebase
     }
 
     func logIn() {
+        print("[FirebaseAuthenticationManager] logging in")
```
I noticed that the logged in method accesses UI stuff, not sure if there's other ways to do it because this seems to violate a lot of SoC.

keanecjy replied:
Also it seems like its getting runtime warnings here

<img width="1841" alt="image" src="https://user-images.githubusercontent.com/65328119/162601905-a8d6b1b4-e37f-45b5-a4db-5c4e14621e73.png">

Vielheim replied:
I don't think this can be changed because the Google login is the one handling the view.

Vielheim replied:
I don't think this can be changed because the Google login is the one handling the view.

keanecjy reviewed (commented):
Some comments

Commit:
remove unused code

Commit:
remove unused code

Commit:
remove reference of AuthenticationManager from Service layer
- The services reference an authentication service instead

Change FirebaseAuthenticationDelegate to AuthenticationManagerDelegate
- Should not be referencing a firebase instance

Added AuthenticatedUser struct to wrap around Firebase.User
- compatible for other authentication strategies

Commit:
change dependency from authManager to authService in Repo layer
